crates/ra_hir_ty/src/method_resolution.rs

```diff
@@ -378,11 +378,16 @@ fn iterate_trait_method_candidates<T>(
     // if ty is `impl Trait` or `dyn Trait`, the trait doesn't need to be in scope
     let inherent_trait = self_ty.value.inherent_trait().into_iter();
     // if we have `T: Trait` in the param env, the trait doesn't need to be in scope
-    let traits_from_env = env
-        .trait_predicates_for_self_ty(&self_ty.value)
-        .map(|tr| tr.trait_)
-        .flat_map(|t| all_super_traits(db, t));
-    let traits = inherent_trait.chain(traits_from_env).chain(traits_in_scope.iter().copied());
+    let traits = if let Ty::Placeholder(_) = self_ty.value {
+        let traits_from_env = env
+            .trait_predicates_for_self_ty(&self_ty.value)
+            .map(|tr| tr.trait_)
+            .flat_map(|t| all_super_traits(db, t));
+        Box::new(inherent_trait.chain(traits_from_env)) as Box<dyn Iterator<Item = _>>
```

Comment:
`traits_from_env` will usually be tiny, I think it'd be easier to just collect it into a vec and use `Vec::new()` in the else branch

Reply:
Done.